keps/sig-network/2009-dns-autopath/README.md

```diff
@@ -0,0 +1,228 @@
+	   This takes 3 parameters - autopath suffix, cluster suffix and schema version.
+	
+NodeLocal DNSCache or any sidecar that generates searchpaths needs config blocks similar to :
+
+(only relevant plugins included in this example)
+
+```
+< Add any custom stubdomains config here>
+
+ap.k8s.io:53 {
+  cache
+  gensearchpath ap.k8s.io cluster.local v1 {
+    forward <kube-dns service IP>
+    fallthrough
+  }
+  forward . /etc/resolv.conf
+}
+
+cluster.local:53 {
+  cache
+  forward <kube-dns service IP>
+}
+
+:53 {
+  cache
+  forward . /etc/resolv.conf
+}
+```
+
+
+Let us consider a few scenarios and how DNS resolution will work.
+
+1) Client pod looks up - "mytestservice", but did not request searchpath expansion.
+
+ This can happen if user issued a command like 
+ 
+ `dig mytestservice`
+ 
+ This query is expected to fail since it is not fully qualified and the service name does not exist as-such. Query hits the 3rd block above and is looked up by the external DNS server. No EDNS0 options get added.
+
+2) Client pod looks up - "mytestservice" with searchpath expansion. Query is sent as `mytestservice.search.default.cluster.local.ap.k8s.io` and hits the first block. The namespace and suffix are extracted and the EDNS0 option contains "default.svc.cluster.local, svc.cluster.local, cluster.local,<other searchpaths>". Assuming service name is found, no fallthrough is needed.
+
+The same steps apply when a client pod looks up any other service name with part or even all of the FQDN.
+
+3) Client pod looks up "google.com" with searchpath expansion. Query is received as `google.com.search.default.cluster.local.ap.k8s.io`. The searchpath expanded queries have no response, so "google.com" fallthrough to external DNS Server without EDNS0 option.
+
+4) A service name FQDN "mytestservice.default.svc.cluster.local" is looked up, ndots is set to 4.
+No searchpath expansion, query hits the 2nd block "cluster.local" and is forwarded to kube-dns service. 
+5) Queries matching a custom stubDomain will match their own config block and not be subject to this searchpath expansion.
+
+There are 3 parts to the implementation:
+
+1) Add a new CoreDNS plugin to generate searchpaths and attach EDNS0 option to a query. 
+
+2) Enhance autopath plugin to read searchpaths from this new EDNS0 option. 
+
+3) Enhance autopath to also parse the new searchpath format - `search.$NS.$SUFFIX.ap.k8s.io`. Doing this ensures that DNS resolution works even if user does not have a sidecar or node-local-dns pod running.
+
+### Test Plan
+
+The existing DNS conformance tests will be run against clusters with pods using the new "clusterFirstWithAutopath" dnsPolicy.
+
+## Graduation Criteria
+
+In order to graduate to beta, we will need:
+
+Conformance tests exercising this new searchpath.
+Verify performance/scalability via automated testing and measure the improvement over the existing `clusterFirst` mode.
```

Comment:
In some user cases, we want to enable `autopath` in our clusters, however, we still want a benchmark to compare with `enable autopath` and `disable autopath` to make sure that there is no performance decrease after we enable it.